core/new-gui/src/app/common/service/user/user-dictionary/dictionary.service.ts

```diff
@@ -0,0 +1,518 @@
+    result: {type: 'string'}
+  },
+  additionalProperties: false,
+  required: ['code', 'result']
+};
+
+export type UserDictionary = {
+  [Key: string]: string;
+};
+
+export enum EVENT_TYPE {
+  GET,
+  SET,
+  DELETE,
+  GET_ALL
+}
+
+export type GET_EVENT = {
+  type: EVENT_TYPE.GET
+  key: string
+  value: string
+};
+
+export type SET_EVENT = {
+  type: EVENT_TYPE.SET
+  key: string
+  value: string
+};
+
+export type DELETE_EVENT = {
+  type: EVENT_TYPE.DELETE
+  key: string
+};
+
+export type GET_ALL_EVENT = {
+  type: EVENT_TYPE.GET_ALL
+  value: UserDictionary
+};
+
+export type USER_DICT_EVENT = GET_EVENT | SET_EVENT | DELETE_EVENT | GET_ALL_EVENT;
+
+/**
+ * This error is only thrown if attempting to get a dictionary proxy before initialization is complete
+ */
+export class NotReadyError extends Error {
+  constructor(message?: string) {
+    super(message);
+    Object.setPrototypeOf(this, new.target.prototype);
+    this.name = 'NotReadyError';
+  }
+}
+
+@Injectable({
+  providedIn: 'root'
+})
+export class DictionaryService {
+  private static ajv = new Ajv({useDefaults: true});
+  private static validateErrorResponse = DictionaryService.ajv.compile(errorResponseSchema);
+  private static validateEntryResponse = DictionaryService.ajv.compile(entryResponseSchema);
+  private static validateDictResponse = DictionaryService.ajv.compile(dictResponseSchema);
+  private static validateConfirmationResponse = DictionaryService.ajv.compile(confirmationResponseSchema);
+  public static readonly USER_DICTIONARY_ENDPOINT = 'users/dictionary';
+  private dictionaryEventSubject = new Subject<USER_DICT_EVENT>();
+  private localUserDictionary: UserDictionary = {}; // asynchronously initialized after construction (see initLocalDict)
+  private ready: { promise: Promise<boolean>, value: boolean } = {promise: Promise.resolve(false), value: false};
+
+  constructor(private http: HttpClient) {
+    this.initLocalDict();
+    this.handleDictionaryEventStream();
+  }
+
+  /**
+   * get a value from the backend.
+   * Also produces dictionary events and updates dictionary proxy objects as a side effect.
+   * keys and values must be strings.
+   * @param key string key that uniquely identifies a value
+   * @returns string value corresponding to the key from the backend;
+   * throws Error("No such entry") (invalid key) or Error("Invalid session") (not logged in).
+   */
+  public get(key: string): Observable<string> {
+    const url = `${AppSettings.getApiEndpoint()}/${DictionaryService.USER_DICTIONARY_ENDPOINT}/get`;
+    const request: GetEntryRequest = {requestType: GET_REQUEST_TYPE.GET_ENTRY, key: key};
+    const result = this.http.post<UserDictionaryResponse>(url, request)
+      .pipe(
+        map(this.handleEntryResponse.bind(this, key)),
+        catchError(DictionaryService.handleErrorResponse),
+        shareReplay());
+
+    result.subscribe(); // causes post request to actually be sent (see cold observables)
+    return result;
```

Comment:
I am not sure about this observable usage, even if I read about cold observables. Maybe @zuozhiw can give more inputs?